### PR TITLE
Increase test coverage for transaction pool module

### PR DESCRIPTION
**Pull request by dkijania**
Increase coverage for ledger reorg functionality with new property based tests. Apart from general all valid commands generator one can also configure generator to send some edge cases (zkapp which blocking sending further transactions or replace small amount transaction with big amount transaction which should prevent from applying entire chain of commands due to balance drain) . Last commit propose reorganization of all transaction tests since reorg tests require mini test framework.

PR related to https://github.com/MinaProtocol/mina/pull/16401

Tests currently are failing but after we merge above PR they should be green. 

Notes to reviewer: 
- I'm not sure if test reorganization is a good idea (commit e38d28d6a04e63124f1475ee97412d5bc7ee3d0a), no problem with reverting this to commit 9cdcb4d56af5855d7f5f406f684ae1748f1bb485). My reasoning was that I added a quite big chunk of code and thought that maybe we should split tests for better readibility. There is already network_pool/tests folder which contains alco tests. Maybe we can put all of them there too ? 
- I've done one uber property test, which helped me understand how reorg is working. I'm wondering if we should add more fine grain tests which could be used also documentation for reorg mechanism
- I had a big troubles to nicely modify already built User_command (adding nonces or create mirror transactions in other branch), so i decided to work on specification on minor/major sequences first and then create commands based on them.

### CODE REVIEW

**Comment by dkijania** [comment]
!ci-build-me

**Comment by georgeee** [comment]
I like the idea of extraction of transaction pool tests to a standalone file. Let's restructure the PR in the following way:

1. Create a separate predecessor PR that will move the test code out of `transaction_pool.ml` to `test/` subdirectory
2. Retain in this PR only what's related to **new** tests

@dkijania

File: src/lib/network_pool/test_utils.ml [hunk 1]
```diff
@@ -0,0 +1,831 @@
...
+    let apply_initial_ledger_state t init_ledger_state =
+      let new_ledger =
+        Mina_ledger.Ledger.create_ephemeral
+          ~depth:(Mina_ledger.Ledger.depth !(t.best_tip_ref))
+          ()
+      in
+      Mina_ledger.Ledger.apply_initial_ledger_state new_ledger init_ledger_state ;
+      t.best_tip_ref := new_ledger
+
+  let ledger_snapshot t =
```
Comment (georgeee) [inline src/lib/network_pool/test_utils.ml:168]: Suggest to inline it into `src/lib/network_pool/transacation_pool_reorg_tests.ml` and return `Account_spec.t` instead of a tuple

File: src/lib/network_pool/transacation_pool_reorg_tests.ml [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+    let (kp,balance,nonce,_)  = ledger_row in
+    create 
+       (find_in_test_keys kp) 
+       (Currency.Amount.to_nanomina_int balance)
+       (Unsigned.UInt32.to_int nonce)
+  
+  let apply_payment amount fee t = 
+    create t.key_idx (t.balance - amount - fee) (t.nonce + 1)
+
+  let apply_zkapp fee t = create t.key_idx (t.balance - fee) (t.nonce + 1)
```
Comment (georgeee) [inline src/lib/network_pool/transacation_pool_reorg_tests.ml:41]: Are all zkapps in this test created with with zero account updates?

If so, this looks ok, but let's add a comment in the code about this matter.

File: src/lib/network_pool/transacation_pool_reorg_tests.ml [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+    - The prefix ledger* contains just enough funds to process major sequence, with a small surplus.
+    - When applying *minor sequence* without the transaction `T'` (of the same nonce as the large-amount transaction `T` in major sequence), 
+        the sequence becomes partially applicable, forcing the mempool logic to drop some transactions at the end of *minor sequence*.
+  *)
+
+  (*find account in major and minor brnaches with the same nonces and similar balances (less than 100k mina diff)*)
+  let minor_acc_opt = Array.find_map minor ~f:(fun minor_acc -> 
+      Array.find major ~f:(fun major_acc -> 
+            Int.equal minor_acc.nonce major_acc.nonce &&
+            minor_acc.balance - major_acc.balance < 100_000_000_000
```
Comment (georgeee) [inline src/lib/network_pool/transacation_pool_reorg_tests.ml:152]: What is the chance that such two accounts won't be found?

Won't the probability of the test being executed be too low?

**Review by georgeee — CHANGES_REQUESTED** [review]
_Review is still in progress, posting the first few comments..._

**Review by georgeee — CHANGES_REQUESTED** [review]

File: src/lib/network_pool/transacation_pool_reorg_tests.ml [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+        but after the permission-modifying transaction in major sequence, 
+        the new transaction becomes invalid and must be dropped.
+  *)
+  let%bind permission_change_cmd = Command_spec.gen_zkapp_blocking_send major in
+  let sender = Command_spec.sender permission_change_cmd in
+  (* We need to increase nonce so transaction has a chance to be placed in the pool.
+    Otherwise it will be dropped as we already have transaction with the same nonce from major sequence
+  *)
+  let sender = major.(sender.key_idx) in
+  let%bind aux_minor_cmd = Command_spec.gen_single_from minor (sender.key_idx,sender) in
```
Comment (georgeee) [inline src/lib/network_pool/transacation_pool_reorg_tests.ml:203]: this logic won't imitate the required behavior if `maj1 >= min1`, where `maj1` and `min1` represent the number of txs from `sender` in `major` and `minor` respectively.

I.e. `aux_minor_cmd` will simply be thrown away. We need to add as many aux cmds as necessary to ensure that resulting minor sequence has more transactions from sender than major.

File: src/lib/network_pool/transacation_pool_reorg_tests.ml [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+    - In major sequence, a transaction modifies an account's permissions:
+        1. It removes the permission to maintain the nonce.
+        2. It removes the permission to send transactions.
+    - In minor sequence, there is a regular transaction involving the same account, 
+        but after the permission-modifying transaction in major sequence, 
+        the new transaction becomes invalid and must be dropped.
+  *)
+  let%bind permission_change_cmd = Command_spec.gen_zkapp_blocking_send major in
+  let sender = Command_spec.sender permission_change_cmd in
+  (* We need to increase nonce so transaction has a chance to be placed in the pool.
```
Comment (georgeee) [inline src/lib/network_pool/transacation_pool_reorg_tests.ml:199]: where is the actual increase happening?

File: src/lib/network_pool/transacation_pool_reorg_tests.ml [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+      sender = account_spec
+      ; fee = minimum_fee
+    }) 
+
+  let gen_single_from (spec:Account_spec.t array) (idx,account_spec) = 
+      let open Quickcheck.Generator.Let_syntax in
+        let%bind receiver_idx = test_keys |> Array.mapi ~f:(fun i _-> i) |> Quickcheck_lib.of_array in
+        let%bind amount = Int.gen_incl 5_000_000_000_000 10_000_000_000_000  in
+        let new_account_spec = Account_spec.apply_payment amount minimum_fee account_spec in 
+        Array.set spec idx new_account_spec;
```
Comment (georgeee) [inline src/lib/network_pool/transacation_pool_reorg_tests.ml:77]: What if balance goes negative? Perhaps we set amount to `0` if amount is less than balance. And if balance is less than min fee, throw an exception.

Comment (georgeee) [reply src/lib/network_pool/transacation_pool_reorg_tests.ml:77]: Case of fee applies to zkapp tx above as well

File: src/lib/network_pool/transacation_pool_reorg_tests.ml [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+          let log_pool_content = List.map pool_state ~f:(fun (fee_payer_pk,nonce) -> 
+            `String ( Printf.sprintf !"%{sexp: Public_key.Compressed.t} : %d" fee_payer_pk nonce )
+          ) in
+          
+          [%log info] "Pool state" ~metadata:[
+            ("pool state", `List log_pool_content )
+          ];
+
+          let assert_pool_contains pool_state (pk,nonce)= 
+            let actual_opt = List.find pool_state ~f:(fun (fee_payer_pk, actual_nonce ) ->
```
Comment (georgeee) [inline src/lib/network_pool/transacation_pool_reorg_tests.ml:330]: Rewrite this and next function: extract calculation of `actual_opt` into a helper function

File: src/lib/network_pool/transacation_pool_reorg_tests.ml [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+                Int.equal acc.key_idx sender.key_idx && 
+                Int.(>) acc.nonce sender.nonce)
+            
+            in  
+            let account_spec = Array.find major_account_spec ~f:(fun spec -> 
+                Int.equal sender.key_idx spec.key_idx && spec.nonce > 0
+              )
+            in
+            match account_spec with
+              | Some account_spec ->
```
Comment (georgeee) [inline src/lib/network_pool/transacation_pool_reorg_tests.ml:368]: Extract `sent_blocking_zkapp`, `total_cost` out of match construction (put it before).

Rewrite `match` without ifs:

```
| Some account_spec when sender.nonce < account_spec.nonce -> (..)
| Some account_spec when sent_blocking_zkapp major_specs pk -> (..)
| Some account_spec when account_spec.balance > total_cost  -> (..)
| Some account_spec -> (..)
| None -> (..)
```

**Review by georgeee — CHANGES_REQUESTED** [review]

File: src/lib/network_pool/transacation_pool_reorg_tests.ml (hunk 2 above)
Comment (dkijania) [reply src/lib/network_pool/transacation_pool_reorg_tests.ml:41]: Maybe name is misleading. apply_zkapp function in context of account_spec is only modify spec state which we are using for tracking balance and nonce. So no zkapp command or update is done. Just noting that from this account zkapp was sent and we are deducting fee and increases nonce

File: src/lib/network_pool/transacation_pool_reorg_tests.ml [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+        
+        [%log info] "Sequences" ~metadata:[
+          ("prefix" , `List log_prefix)
+          ; ("major" , `List log_major)
+          ; ("minor" , `List log_minor)
+          ; ("minor accounts state",`List log_minor_accounts_state )
+          ; ("major accounts state", `List log_major_accounts_state)
+        ];
+
+        let prefix = gen_commands_from_specs (Array.concat [prefix_specs]) test in
```
[the remark this answers is not in the document]
Comment (dkijania) [reply src/lib/network_pool/transacation_pool_reorg_tests.ml:298]: I didn't saw it. thanks!

File: src/lib/network_pool/transacation_pool_reorg_tests.ml [hunk 10]
```diff
@@ -17,7 +17,7 @@ module Account_spec = struct
     ; nonce: int
   } [@@deriving sexp]
 
-  let rec find_in_test_keys ?(n=0) (x:Keypair.t) =
+  let find_in_test_keys (x : Keypair.t) =  Array.findi_exn test_keys ~f:(fun _ -> Keypair.equal x)  |> fst
```
Comment (georgeee) [inline src/lib/network_pool/transacation_pool_reorg_tests.ml:20]: And remove the two lines below

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Review by mrmr1993 — APPROVED** [review]

**Review by mrmr1993 — CHANGES_REQUESTED** [review]
Partial review; still need to do a full pass. I've also proposed https://github.com/MinaProtocol/mina/pull/16467, which is functionally the same, but with opinionated refactors to help me understand the code better.

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

File: src/lib/network_pool/transaction_pool.ml [hunk 11]
```diff
@@ -3156,4 +3163,744 @@ let%test_module _ =
...
+        test_keys
+        |> Array.filter_mapi ~f:(fun i _ ->
+               if Int.equal i (Simple_account.key_idx target_account) then None
+               else Some i )
+        |> Quickcheck_lib.of_array
+      in
+      let%bind major_sequence_length = Int.gen_incl 2 10 in
+      let%bind minor_sequence_length =
+        let%map minor_sequence_length = Int.gen_incl 2 4 in
+        minor_sequence_length + major_sequence_length + initial_nonce
```
Comment (georgeee) [inline src/lib/network_pool/transaction_pool.ml:3502]: Why are we adding the nonce of account  to minor sequence length? What if account has nonce of `100000` in the initial ledger?

Comment (georgeee) [reply src/lib/network_pool/transaction_pool.ml:3502]: Is it to ensure that extended minor sequence is longer than number of transactions associated with the account in pre-existing major sequence?

File: src/lib/network_pool/transaction_pool.ml [hunk 12]
```diff
@@ -3156,4 +3163,744 @@ let%test_module _ =
...
+      in
+      let initial_balance = Simple_account.balance target_account in
+      let half_initial_balance = Simple_account.balance target_account / 2 in
+      let gen_sequence_and_update_account ledger len =
+        let account = ref (Simple_ledger.get ledger target_account_idx) in
+        let%map sequence =
+          Quickcheck_lib.init_gen_array len ~f:(fun _ ->
+              let%bind amount =
+                Int.gen_incl 5_000_000_000_000_000 (half_initial_balance / len)
```
Comment (georgeee) [inline src/lib/network_pool/transaction_pool.ml:3511]: What if `(half_initial_balance / len) < 5_000_000_000_000_000`? will it fail with an exception?

Comment (georgeee) [reply src/lib/network_pool/transaction_pool.ml:3511]: So, I just checked that initial balance is always 900mln. I still see little benefit in hardcoding the `5mln` amount here

File: src/lib/network_pool/transaction_pool.ml [hunk 13]
```diff
@@ -3156,4 +3163,744 @@ let%test_module _ =
...
+                (Simple_account.key_idx sender_on_minor)
+            in
+            Simple_command.gen_single_and_update_ledger minor
+              (sender_on_minor_idx, sender_on_minor) )
+      in
+      return
+        { prefix_commands
+        ; major_commands =
+            Array.append major_commands [| permission_change_cmd |]
+        ; minor_commands = Array.append minor_commands aux_minor_cmd
```
Comment (georgeee) [inline src/lib/network_pool/transaction_pool.ml:3645]: Is it possible to mix aux commands (which are expected to be dropped) with the regular commands from `minor_commands`? Or just generate some new random commands and run `gen_merge` on this new sequence and `aux_commands`. We want to ensure that no valid tx from other account is dropped together with aux commands that need to be dropped.

**Review by georgeee — APPROVED** [review]
I like the current test implementation, very readable and precise!

I left more than a dozen review comments, but I doubt any of them prevents the test from doing its job alright in the current setting.

I still want these changes to be applied, mostly to make the test more robust (so that it would work practically on any initial ledger), slightly more readable and to test a slightly fuller distribution of cases. But I'm quite fine with applying these changes in a follow-up PR

File: src/lib/network_pool/transaction_pool.ml (hunk 11 above)
Comment (dkijania) [reply src/lib/network_pool/transaction_pool.ml:3501]: Exactly, we need to take into consideration case in which major branch has more transactions for particular account. Therefore on minor we need to generate more transaction to keep up with nonce

File: src/lib/network_pool/transaction_pool.ml (hunk 12 above)
Comment (dkijania) [reply src/lib/network_pool/transaction_pool.ml:3510]: hm, so what should be the minimal amount ? should it be related to half_initial_balance?

**Comment by dkijania** [comment]
!ci-build-me

File: src/lib/network_pool/transaction_pool.ml [hunk 14]
```diff
@@ -3828,78 +3861,64 @@ let%test_module _ =
...
-                         should be in the pool"
-                        ~metadata:
-                          [ ( "sent from"
-                            , `String
-                                (Printf.sprintf
-                                   !"%{sexp: Public_key.Compressed.t} -> %d"
-                                   pk nonce ) )
-                          ] ;
-                      assert_pool_contains pool_state (pk, nonce) ) ;
+                  if sender.nonce < Simple_account.nonce account_spec then (
```
Comment (dkijania) [inline src/lib/network_pool/transaction_pool.ml:3867]: after changing find_by_key_idx i had to modify match back to if-else

File: src/lib/network_pool/transaction_pool.ml [hunk 15]
```diff
@@ -3502,6 +3512,20 @@ let%test_module _ =
       in
       let initial_balance = Simple_account.balance target_account in
       let half_initial_balance = Simple_account.balance target_account / 2 in
+      let recieved_amount =
```
Comment (dkijania) [inline src/lib/network_pool/transaction_pool.ml:3515]: Another safe guard to make sure we deduct correct amount from account balance

File: src/lib/network_pool/transaction_pool.ml (hunk 12 above)
Comment (georgeee) [reply src/lib/network_pool/transaction_pool.ml:3534]: yeah, let's make it proportional to `half_initial_balance/len`

**Comment by dkijania** [comment]
!ci-build-me

**Review by mrmr1993 — APPROVED** [review]
